src/farmer.rs

```diff
@@ -93,20 +119,91 @@ impl PlotDescription {
 }
 
 mod builder {
+    use crate::generate_builder;
     use libp2p_core::Multiaddr;
 
     /// Technical type which stores all
-    #[derive(Debug, derive_builder::Builder)]
+    #[derive(Debug, Clone, derivative::Derivative, derive_builder::Builder)]
     #[builder(pattern = "owned", build_fn(name = "_build"), name = "Builder")]
     pub struct Configuration {
+        /// DSN options
+        #[builder(default, setter(into))]
+        pub dsn: Dsn,
+    }
+
+    /// Farmer DSN
+    #[derive(Debug, Clone, Default, derive_builder::Builder)]
+    #[builder(pattern = "owned", build_fn(name = "_build"), name = "DsnBuilder")]
+    pub struct Dsn {
         /// Listen on
-        #[builder(default = "vec![]")]
-        #[builder(setter(strip_option))]
+        #[builder(default)]
         pub listen_on: Vec<Multiaddr>,
         /// Bootstrap nodes
-        #[builder(setter(strip_option))]
-        #[builder(default = "vec![]")]
+        #[builder(default)]
         pub bootstrap_nodes: Vec<Multiaddr>,
+        /// Determines whether we allow keeping non-global (private, shared, loopback..) addresses in Kademlia DHT.
+        #[builder(default)]
+        pub allow_non_global_addresses_in_dht: bool,
+    }
+
+    generate_builder!(Dsn);
+}
+
+impl builder::Dsn {
+    async fn configure_dsn(
```

Comment:
maybe add docs here?

Reply:
It is an internal thing, so I guess we don't need to document it.